src/pudl/output/eia923.py

```diff
@@ -156,9 +159,79 @@ def generation_fuel_eia923(pudl_engine, freq=None,
     return out_df
 
 
-def fuel_receipts_costs_eia923(pudl_engine, freq=None,
-                               start_date=None, end_date=None,
-                               fill=False, roll=False):
+def generation_fuel_all_eia923(gf: pd.DataFrame, gfn: pd.DataFrame) -> pd.DataFrame:
+    """
+    Combine nuclear and non-nuclear generation fuel tables into a single output.
+
+    The nuclear and non-nuclear generation fuel data are reported at different
+    granularities. For non-nuclear generation, each row is a unique combination of date,
+    plant ID, prime mover, and fuel type. Nuclear generation is additionally split out
+    by nuclear_unit_id (which happens to be the same as generator_id).
+
+    This function aggregates the nuclear data across all nuclear units within a plant so
+    that it is structurally the same as the non-nuclear data and can be treated
+    identically in subsequent analyses. Then the nuclear and non-nuclear data are
+    concatenated into a single dataframe and returned.
+
+    Args:
+        gf: non-nuclear generation fuel dataframe.
+        gfn: nuclear generation fuel dataframe.
+
+    """
+    primary_key = [
+        "report_date",
+        "plant_id_eia",
+        "prime_mover_code",
+        "energy_source_code",
+    ]
+    sum_cols = [
+        "fuel_consumed_for_electricity_mmbtu",
+        "fuel_consumed_for_electricity_units",
+        "fuel_consumed_mmbtu",
+        "fuel_consumed_units",
+        "net_generation_mwh",
+    ]
+    replace_cols = ["nuclear_unit_id", "fuel_mmbtu_per_unit"]
+    # Rather than enumerating all of the non-data columns, identify them by process of
+    # elimination, in case they change in the future.
+    non_data_cols = list(
+        set(gfn.columns) - set(primary_key + sum_cols + replace_cols)
+    )
+
+    gfn_gb = gfn.groupby(primary_key)
+    # Ensure that all non-data columns are homogeneous within groups
+    if not (gfn_gb[non_data_cols].nunique() == 1).all(axis=None):
+        raise ValueError(
+            "Found inhomogeneous non-data cols while aggregating nuclear generation."
+        )
+    agg_dict = dict(
+        **{col: "first" for col in non_data_cols},
+        **{col: "sum" for col in sum_cols},
+    )
+    gfn_agg = (
+        gfn_gb.agg(agg_dict)
+        .assign(fuel_mmbtu_per_unit=lambda x: x.fuel_consumed_mmbtu / x.fuel_consumed_units)
+        .reset_index()
+    )
+    # Nuclear plants don't report units of fuel consumed, so heat rate is infinite.
+    # However, some nuclear plants report using small amounts of DFO. Replace the
+    # infinite heat rates with NA:
+    gfn_agg.loc[np.isinf(gfn_agg.fuel_mmbtu_per_unit), "fuel_mmbtu_per_unit"] = np.nan
```

Comment:
Not sure I understand what you're getting at here. Infinity is non-null, so wouldn't this would have no impact on the infinite values, which are what I'm trying to replace? And then set all the null values to... `np.nan` which would do nothing?
```py
np.where(x.fuel_consumed_mmbtu.notnull(), x.fuel_consumed_mmbtu / x.fuel_consumed_units, np.na)
```
set a

Reply:
oh i see. well I think I'd still do `np.where(x.fuel_consumed_mmbtu == 0` in that case. but it is not a big deal either way.

Reply:
also are they 0's bc of the groupby?? i'd add `min_count=1` into the sum so you don't get a bunch of zeros.

Reply:
No, they're zeroes in the DB.